ambry-tools/src/main/java/com/github/ambry/tools/perf/serverperf/ServerPerfNetworkQueue.java

```diff
@@ -0,0 +1,256 @@
+  private final int pollTimeout;
+  private final int maxParallelism;
+
+  private final Semaphore maxParallelRequest;
+  private final ExecutorService executorService;
+
+  private final CountDownLatch shutDownLatch;
+
+  private final Time time;
+  private boolean isShutDown;
+
+  private final int operationsTimeOutMs;
+
+  private final Map<Integer, Long> pendingCorrelationIdToStartTimeMs = new LinkedHashMap<>();
+  private final Map<Integer, Integer> pendingCorrelationIdToNetworkClientIdx = new HashMap<>();
+  private final Map<Integer, RequestInfo> pendingCorrelationIdToRequestInfo = new HashMap<>();
+
+  private static final Logger logger = LoggerFactory.getLogger(ServerPerfNetworkQueue.class);
+
+  /**
+   *
+   * @param verifiableProperties properties
+   * @param metrics http2 client metrics
+   * @param time  ambry Time
+   * @param maxParallelism maximum parallel requests that can be processed
+   * @param clientCount total number of network clients to use
+   * @param operationsTimeOutSec time for a request to mark as timed out
+   * @throws Exception exception
+   */
+  ServerPerfNetworkQueue(VerifiableProperties verifiableProperties, Http2ClientMetrics metrics, Time time,
+      int maxParallelism, int clientCount, int operationsTimeOutSec) throws Exception {
+    SSLFactory sslFactory = new NettySslHttp2Factory(new SSLConfig(verifiableProperties));
+    Http2ClientConfig http2ClientConfig = new Http2ClientConfig(verifiableProperties);
+    Http2NetworkClientFactory networkClientFactory =
+        new Http2NetworkClientFactory(metrics, http2ClientConfig, sslFactory, time);
+    networkClients = new ArrayList<>();
+    for (int i = 0; i < clientCount; i++) {
+      networkClients.add(networkClientFactory.getNetworkClient());
+    }
+    this.time = time;
+    isShutDown = false;
+    clientIndex = 0;
+    requestInfos = new ConcurrentLinkedQueue<>();
+    responseInfos = new ConcurrentLinkedQueue<>();
+    pollTimeout = 0;
+    this.operationsTimeOutMs = operationsTimeOutSec * 1000;
+    this.maxParallelism = maxParallelism;
+    maxParallelRequest = new Semaphore(maxParallelism, true);
+    executorService = Executors.newFixedThreadPool(maxParallelism);
+    shutDownLatch = new CountDownLatch(1);
+  }
+
+  /**
+   * Adds the request info to the queue and ready for polling
+   * @param requestInfo request info
+   * @throws Exception
+   */
+  void submit(RequestInfo requestInfo) throws Exception {
+    if (isShutDown) {
+      throw new ShutDownException();
+    }
+    maxParallelRequest.acquire();
+    if (isShutDown) {
+      throw new ShutDownException();
+    }
+    requestInfos.offer(requestInfo);
+  }
+
+  /**
+   * Checks and processes if any responses are in the queue
+   * @param responseInfoProcessor processor
+   * @throws Exception
+   */
+  void poll(ResponseInfoProcessor responseInfoProcessor) throws Exception {
+    ResponseInfo responseInfo = responseInfos.poll();
```

Comment:
Put this under if condition below, so we won't miss the last response info.

Reply:
Changed it.